server/src/main/java/io/druid/server/QueryManager.java

```diff
@@ -24,23 +24,36 @@
 import com.google.common.collect.SetMultimap;
 import com.google.common.util.concurrent.ListenableFuture;
 import com.google.common.util.concurrent.MoreExecutors;
+import com.metamx.common.IAE;
+import io.druid.query.DataSource;
 import io.druid.query.Query;
+import io.druid.query.QueryDataSource;
 import io.druid.query.QueryWatcher;
+import io.druid.query.TableDataSource;
+import io.druid.query.UnionDataSource;
 
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 
 public class QueryManager implements QueryWatcher
 {
+
   final SetMultimap<String, ListenableFuture> queries;
+  final Map<String, List<String>> queryDatasources;
 
   public QueryManager()
   {
     this.queries = Multimaps.synchronizedSetMultimap(
         HashMultimap.<String, ListenableFuture>create()
     );
+    this.queryDatasources = new ConcurrentHashMap<>();
```

Comment:
This breaks compat with the way query futures are handled. Specifically, the queries map can have multiple futures added, whereas this cannot. As such even though `this.queries` may have the same queryId for widely varying actual queries, there is no methodology in `this.queryDatasources` to distinguish them.

This means I can hammer some queries with the same query ID to try and achieve a race condition where the result of `QueryManager. getQueryDatasources` or `QueryManager. getDatasources` is non deterministic.